libp2p/module.go

```diff
@@ -0,0 +1,367 @@
+// the given stream for handling at the network level. Used for handling "direct"
+// messages (i.e. one specific target node).
+func (mod *libp2pModule) readStream(stream libp2pNetwork.Stream) {
+	closeStream := func() {
+		if err := stream.Close(); err != nil {
+			mod.logger.Error().Err(err)
+		}
+	}
+
+	// NB: time out if no data is sent to free resources.
+	if err := stream.SetReadDeadline(newReadStreamDeadline()); err != nil {
+		mod.logger.Error().Err(err).Msg("setting stream read deadline")
+		// TODO: abort if we can't set a read deadline?
+	}
+
+	data, err := io.ReadAll(stream)
+	if err != nil {
+		mod.logger.Error().Err(err).Msg("reading from stream")
+		closeStream()
+		// NB: abort this goroutine
+		// TODO: signal this somewhere?
+		return
+	}
+	defer closeStream()
+
+	mod.handleNetworkData(data)
+}
+
+// readFromSubscription is intended to be called in a goroutine. It continuously
+// reads from the subscribed topic in preparation for handling at the network level.
+// Used for handling "broadcast" messages (i.e. no specific target node).
+func (mod *libp2pModule) readFromSubscription(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			msg, err := mod.subscription.Next(ctx)
+			if err != nil {
+				mod.logger.Error().Err(err).Msg("reading from subscription")
+			}
+
+			// NB: ignore messages from self
+			if msg.ReceivedFrom == mod.host.ID() {
+				continue
+			}
+
+			mod.handleNetworkData(msg.Data)
+		}
+	}
+}
+
+func (mod *libp2pModule) handleNetworkData(data []byte) {
+	appMsgData, err := mod.network.HandleNetworkData(data)
+	if err != nil {
+		mod.logger.Error().Err(err).Msg("handling network data")
+		return
+	}
+
+	// There was no error, but we don't need to forward this to the app-specific bus.
+	// For example, the message has already been handled by the application.
+	if appMsgData == nil {
+		return
+	}
+
+	networkMessage := messaging.PocketEnvelope{}
+	if err := proto.Unmarshal(appMsgData, &networkMessage); err != nil {
+		mod.logger.Error().Err(err).Msg("Error decoding network message")
+		return
+	}
+
+	event := messaging.PocketEnvelope{
+		Content: networkMessage.Content,
+	}
+
+	mod.GetBus().PublishEventToBus(&event)
+}
+
+// getMultiaddr returns a multiaddr constructed from the `hostname` and `port`
+// in the P2P config which pas provided upon creation.
+func (mod *libp2pModule) getMultiaddr() (multiaddr.Multiaddr, error) {
+	// TECHDEBT: as soon as we add support for multiple transports
+	// (i.e. not just TCP), we'll need to do something else.
+	return network.Libp2pMultiaddrFromServiceUrl(fmt.Sprintf(
+		"%s:%d", mod.cfg.Hostname, mod.cfg.Port,
+	))
+}
+
+// newReadStreamDeadline returns a future deadline
+// based on the read stream timeout duration.
+func newReadStreamDeadline() time.Time {
+	return time.Now().Add(readStreamTimeoutDuration)
+}
+
+func (mod *libp2pModule) HandleEvent(msg *anypb.Any) error {
```

Comment:
Would you mind grouping the Exported methods (Capitalized) first and then the unexported ones (lowercased)?

I guess that it's just `HandleEvent` that needs to be moved up, above `handleStream`

Reply:
Yes, of course - thanks for calling this out! :smile: